File: Python/Data Structures & Algorithms Module - Python/Week4/ArrayStack.py
"""
ArrayStack.py
Billy Cussen
12/11/2020
"""
from Node import Node

class ArrayStack:

    def __init__(self):
        self.top = -1
        self.size = 2
        self.stack = []

    def isEmpty(self):
        return self.top == -1

    def push(self, elem):
        if self.top == self.size - 1:
            self.reallocate()
        self.top+=1
        self.stack.insert(0,elem)

    def reallocate(self):
        self.size*=2

    def pop(self):
        if self.isEmpty():
            raise Exception("Stack is Empty, cannot pop!")
        else:
            self.top-=1
            return self.stack[0]

    def peek(self):
        if self.isEmpty():
            raise Exception("Stack is Empty, cannot peek!")
        else:
            return self.stack[0]

    def mySize(self):
        return self.sizeHelper(self.stack, self.top)

    def sizeHelper(self, current, index):
        if(index==-1):
            return 0
        index-=1
        return 1 + self.sizeHelper(self.stack, index)

    def toString(self):
        if self.isEmpty():
            raise Exception("Can't traverse, Stack is Empty!")
        result = "***ARRAY STACK TRAVERSAL\n"
        position = 0
        index = self.mySize()-1
        return self.toStringHelper(self.stack, result, position,index)

    def toStringHelper(self, stack, result, position, index):
        if(index==-1):
            result+="***END***\n"
            return result
        result+="Position "+str(position)+", Element "+str(stack[index])+"\n"
        position += 1
        index -= 1
        return self.toStringHelper(stack, result, position, index)

File: Java/Data Structures & Algorithms Module - Java/Revision/Stack/ArrayStack.java
/*
ArrayStack.java
Billy Cussen
07/01/2021
*/
import java.util.Arrays;

public class ArrayStack<T> {
    
    int top;
    int size;
    T[] stack;

    public ArrayStack(int size){
        this.size = size;
        this.top = -1;
        this.stack = (T[]) new Object[this.size];
    }

    public int size(){
        return sizeHelper(this.stack,top);
    }

    private int sizeHelper(T[] items, int top){
        if(items[top]==null){
            return 0;
        }
        return 1+sizeHelper(items, ++top);
    }

    public boolean isEmpty(){
        return top == -1;
    }

    @Override
    public String toString(){
        int counter=0, index=this.top;
        String result = "***ARRAYSTACK TRAVERSAL***\n";
        while(index>=0){
            result+="Position "+counter+", Element: "+this.stack[index]+"\n";
            index--;
            counter++;
        }
        result+="***END***\n";
        return result;
    }

    public void push(T element){
        if(this.top == this.size-1){
            reallocate();
        }
        this.top+=1;
        this.stack[this.top] = element;
    }

    private void reallocate(){
        this.size*=2;
        this.stack = Arrays.copyOf(this.stack, this.size);
    }

    public T pop(){
        if(isEmpty()){
            throw new NullPointerException("Can't pop, Stack is Empty!");
        } else{
            T element = this.stack[this.top];
            this.top--;
            return element;
        }
    }

    public T peek(){
        if(isEmpty()){
            throw new NullPointerException("Can't peek, Stack is Empty!");
        } else{
            return this.stack[this.top];
        }
    }

}
